# SeqPanther

SeqPanther tool consists set of commands as follows:

- **seqpatcher**: integrates sanger sequencing of missing regions of an incomplete assembly to the assembly. This command is a modification of the [SeqPatcher tool](https://github.com/krisp-kwazulu-natal/seqPatcher).

- **codoncounter**: performs variant calling and generates nucleotide stats at variant sites and reports the impacts of nucleotide changes on amino acids in the translated proteins.

- **transform**: performs transformation of codon counter output in a format where the user can select a variant to integrate into the reference or assemblies using **nusub** command. The transformation can be performed after the removal of undesired changes in files generated by the codon counter or the first transformation is performed and undesired changes are removed from transformed data.

- **nucsub**: integrates the alteration list generated by **codoncounter** in a given reference or assembled genome based on the user recommendations. <!--Providing options to users to select changes of their interests-->

## Operating system compatibility

Unix platforms.

## Dependencies

The tool relies on multiple external programs and python modules as listed below:

### External Tools

1. Samtools

   - For sorting bam files and indexing them.
   - `conda install samtools` to install.

2. Bcftools

   - For parsing variant calling and generating consensus sequences.
   - `conda install bcftools` to install.

3. Muscle (v. 3.8.31)

   - To perform multiple sequence alignment.
   - `conda install muscle=3.8.31` to install.

4. BLAT

   - To query sequence location in the genome.
   - `conda install blat` to install.

5. MAFFT

   - To align consensus sequence against the reference.
   - `conda install mafft` to install.

# Installation

## Manual installation

1. `git clone https://github.com/codemeleon/seqPanther`
2. `cd seqPanther`
3. `python setup.py install`

## Directly from the GitHub repo

`pip install git+https://github.com/codemeleon/seqPanther.git`

# Usages

seqPanther contains four commands, which can be accessed using `seqpanther --help` or just `seqpanther`:

## CodonCounter

This command help is accessible using `seqpanther codoncounter` or `seqpanther codoncounter --help`.

## SeqIn

This command help is accessible using `seqpanther nucsub` or `seqpanther nucsub --help`.

## SeqPatcher

This command help is accessible at `seqpanther seqpatcher` or `seqpanther seqpatcher --help`.

# Example

1. BAM files for five South African Sars-cov-2 samples can be downloaded from [SeqPanther zenodo](https://zenodo.org/record/7601513) page. BAM files were generated by mapping reads mapped against reference (NC_045512.2) using [BWA](https://github.com/lh3/bwa). BWA generated SAM files were converted to BAM and sorted using samtools.
2. The reference sequences fasta and genome annotation GFF files can be downloaded from [NCBI Genome Fasta](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/858/895/GCF_009858895.2_ASM985889v3/GCF_009858895.2_ASM985889v3_genomic.fna.gz) and [ NCBI Genome GFF](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/858/895/GCF_009858895.2_ASM985889v3/GCF_009858895.2_ASM985889v3_genomic.gff.gz) respectively. Downloaded files need to un-archived.
3. Consensus sequences from bam files genrated using `bcftools mpileup -f GCF_009858895.2_ASM985889v3_genomic.fna bam/K032258-consensus_alignment_sorted.bam| bcftools call -c --ploidy 1 | vcfutils.pl vcf2fq > K032258-consensus_alignment_sorted.fastq` and `python fastq2fasta.py -i K032258-consensus_alignment_sorted.fastq -o consensus/K032258-consensus_alignment_sorted.fasta ` to generates fasta file. `fastq2fasta.py` can be downloaded from the project repo.

4. Store downloaded BAM files in a folder, i.e. called `bam`.
5. Run codoncounter on the BAM files using `seqpanther codoncounter -bam bam -rid NC_045512.2 -ref GCF_009858895.2_ASM985889v3_genomic.fna -gff GCF_009858895.2_ASM985889v3_genomic.gff -coor_range 21000-25000` for all the files in bam folder or `seqpanther codoncounter -bam ./bam/K032282-consensus_alignment_sorted.bam -rid NC_045512.2 -ref GCF_009858895.2_ASM985889v3_genomic.fna -gff GCF_009858895.2_ASM985889v3_genomic.gff -coor_range 21000-25000` specific bam file in the folder. The example genomic range `21000-25000` is for S-gene. The command generate three outputs in the current folder: `sub_output.csv` containing substitution events, `indel_output.csv` containing indel events, `codon_output.csv` details on affected amino acids and `output.pdf` of plots of read distribution, location of alteration and type of alteration.
6. Outputs can be explored using a text file reader and pdf reader.
7. To convert changes in integrable in format `seqpanther cc2ns -s sub_output.csv -i indel_output.csv -o changes`. It generates a CSV file for each sample in `./change` folder. The user needs to explore those files and remove the changes they would like not to be integrated.
8. Finally, `seqpanther nucsubs -i NC_045512.2 -r NC_045512.2.fasta -c consensus -t changes -o results` need to be executed to integrate relevant changes to the consensus sequences. The output will be stored in `results` folder.
9. Example data for `seqpanther seqpatcher` is provded in example folder of this project.
10. To run `seqpanther seqpatcher` on the example data, use `seqpanther seqpatcher -s examples/seqpatcher/ab1 -a examples/seqpatcher/assemblies -o Results -t mmf.csv -O sanger.fasta -g 10 -3 True -x del`. It will generate Sanger sequences integrated consensus files in Result folder, a mmf.csv file reporting that sanger data were paired or single ab1, or in single fasta, and sanger.fasta containing all sanger sequences.

# Features

`codoncounter` and `nucsub` commands were specifically defined for this program. However, each sub-command can be installed and used independently of this program.

# Warning

Recursive use of newly generated consensus sequences might result in incorrect final sequence due to the integration indel events.

# Bug reporting

For bug reporting, please open an issue.

# Citation

If you use this software please cite:

SeqPanther: Sequence manipulation and mutation statistics toolset. James Emmanuel San, Stephanie Van Wyk, Houriiyah Tegally, Simeon Eche, Eduan Wilkinson, Aquillah M. Kanzi, Tulio de Oliveira, Anmol M. Kiran. bioRxiv 2023. doi: [10.1101/2023.01.26.525629](https://doi.org/10.1101/2023.01.26.525629)

Zenado:

[![DOI](https://zenodo.org/badge/529166378.svg)](https://zenodo.org/badge/latestdoi/529166378)
